### DISCUSSION TITLE
Add IsolationMode.InstancePerRootTest ?

### DISCUSSION THREAD

**Discussion by sksamuel** [kotest/kotest#4057] [6.0 changes]
It seems to me that InstancePerLeaf is not really what you want, and that InstancePerRootTest would be better. That way it creates a new test instantiation for each top level test, and things nested in there can share data. So kinda like how junit doesi t.

**Comment by bliessens** [comment]
With InstancePerLeaf you must reset (clear) mocks after every single test. It's consistent, regardless of how you (not) nest tests.
With InstancePerLeaf that behaviour is less consistent. Not sure that is a better option.

I have been using Kotest for quite some time without explicitly knowing about this IsolationMode and I do not think that is bad. It's a more advanced feature which you do not need to be exposed to for simple cases. I'm affraid changing the default IsolationMode will have too much impact here.

**Reply by sksamuel** [reply]
Would not change the default.

Just add a new mode, where a new instance is created per top level test, not per final leaf test, which I think makes more sense?

**Comment by jlous** [comment]
I for one will sorely miss InstancePerLeaf.
I use FreeSpec almost exclusively.  

For my functional acceptance tests I prefer SingleInstance, so steps in a scenario can depend on each other without excessive nesting, sort of Cucumber-style.

But for more unit-ish tests, I use InstancePerLeaf:
I want those requirements to be completely independent, and I want the cases run in random order to enforce that they are indeed independent. Like JUnit.
I often bundle cases in root nodes, either to communicate that the cases are related, or because they share certain setup that is not relevant to other cases. The latter can lead to several nesting levels, and I love this possibility, which JUnit lacks.
So in short: my root nodes do not represent the single (independent) test cases. This possibility is to me part of the beauty of FreeSpec and kotest, but it means InstancePerRoot does not do the job.
InstancePerLeaf provides exactly what I need for this: A clean, simple and readable logic where each test receives exactly the history on its path, and nothing else.
Who wouldn't want that?
 
So I really hope a well-defined version of it or something similar can be reintroduced.

Also: I wish the deprecation notice in the docs would link to some sort of explanation of the cited undefined behaviour: under what circumstances does his problem arise, and what are consequences?
